bin/proving-service/src/api/prover.rs

```diff
@@ -0,0 +1,188 @@
+use miden_block_prover::LocalBlockProver;
+use miden_objects::{
+    MIN_PROOF_SECURITY_LEVEL, batch::ProposedBatch, block::ProposedBlock,
+    transaction::TransactionWitness, utils::Serializable,
+};
+use miden_tx::{LocalTransactionProver, TransactionProver};
+use miden_tx_batch_prover::LocalBatchProver;
+use tokio::sync::Mutex;
+use tonic::{Request, Response, Status};
+use tracing::instrument;
+
+use crate::{
+    commands::worker::ProverType,
+    generated::{ProofType, ProvingRequest, ProvingResponse, api_server::Api as ProverApi},
+    utils::MIDEN_PROVING_SERVICE,
+};
+
+/// The provers for the proving service.
+///
+/// This enum is used to store the provers for the proving service.
+/// Only one prover is enabled at a time.
+enum Provers {
+    Transaction(Mutex<LocalTransactionProver>),
+    Batch(Mutex<LocalBatchProver>),
+    Block(Mutex<LocalBlockProver>),
+}
+
+impl Provers {
+    fn new(prover_type: ProverType) -> Self {
+        match prover_type {
+            ProverType::Transaction => {
+                Self::Transaction(Mutex::new(LocalTransactionProver::default()))
+            },
+            ProverType::Batch => {
+                Self::Batch(Mutex::new(LocalBatchProver::new(MIN_PROOF_SECURITY_LEVEL)))
+            },
+            ProverType::Block => {
+                Self::Block(Mutex::new(LocalBlockProver::new(MIN_PROOF_SECURITY_LEVEL)))
+            },
+        }
+    }
+}
+
+pub struct ProverRpcApi {
+    prover: Provers,
+}
+
+impl ProverRpcApi {
+    pub fn new(prover_type: ProverType) -> Self {
+        let prover = Provers::new(prover_type);
+
+        Self { prover }
+    }
+
+    #[instrument(
+        target = MIDEN_PROVING_SERVICE,
+        name = "proving_service:prove_tx",
+        skip_all,
+        ret(level = "debug"),
+        fields(id = tracing::field::Empty),
+        err
+    )]
+    pub fn prove_tx(
```

Comment:
I was going to say that the 3 functions could maybe be collapsed to one, with a `match` that decided the prover within the function, but maybe it's better to keep it the current way to easily decorate the functions with different instrumentation spans